### PR TITLE
Add support for reading /etc/umount.d rather then /etc/umount.conf

### PR DESCRIPTION
**Pull request by rhatdan**
We want to allow container runtimes to ship their own configuration files to specify which mount points they want to umount.  This will eliminate some of the noise in logs.
crio and docker should ship configuration specific to their configuration.

### CODE REVIEW

**Comment by rhatdan** [comment]
@rhvgoyal PTAL

**Comment by rhatdan** [comment]
One left to go.
@rhvgoyal PTAL

**Comment by rhatdan** [comment]
@rhvgoyal Seems to be working pretty well. PTAL.

File: src/oci-umount.c [hunk 1]
```diff
@@ -746,9 +856,9 @@ static int parseBundle(const char *id, yajl_val *node_ptr, char **rootfs, struct
 	config_node = yajl_tree_parse((const char *)configData, errbuf, sizeof(errbuf));
 	if (config_node == NULL) {
 		if (strlen(errbuf)) {
-			pr_perror("parse error: %s: %s: %s", id, config_file_name, errbuf);
+			pr_perror("%s: parse error: %s: %s", id, config_file_name, errbuf);
 		} else {
-			pr_perror("parse error: %s: %s: unknown error", id, config_file_name);
+			pr_perror("%s: parse error: %s: unknown error", id, config_file_name);
```
Comment (rhvgoyal) [inline src/oci-umount.c]: All these changes where we want to change location of container id string, should be in a separate patch. It has nothing to do with the parsing of files.

File: src/oci-umount.c [hunk 2]
```diff
@@ -513,50 +634,39 @@ static int prestart(
...
-
-		if (nr_umounts == MAX_UMOUNTS) {
-			pr_perror("%s: Exceeded maximum number of supported unmounts is %d", id, MAX_UMOUNTS);
+	struct stat stat_buf;
+	if (stat(MOUNTCONF, &stat_buf) == 0) {
+		if (readconf(id, MOUNTCONF, mounts_on_host, &nr_umounts))
+			return -1;
+	} else {
+		if (errno != ENOENT) {
+			pr_perror("%s: Failed to read %s", id, MOUNTCONF);
```
Comment (rhvgoyal) [inline src/oci-umount.c]: So we failed to do stat() on mountconf. Should we say that in error messages. 'Failed to do stat(%s): MOUNTCONF'

File: src/oci-umount.c [hunk 3]
```diff
@@ -513,50 +634,39 @@ static int prestart(
...
-		if (!real_path) {
-			pr_pdebug("%s: Failed to canonicalize path [%s]: %m. Skipping.", id, line);
-			continue;
+	if (stat(ETCMOUNTDIR, &stat_buf) == 0) {
+		if (load_config(id, configs, ETCMOUNTDIR, mounts_on_host, &nr_umounts) < 0) {
+			return EXIT_FAILURE;
 		}
+	} else {
+		if (errno != ENOENT) {
+			pr_perror("%s: Failed to read %s", id, ETCMOUNTDIR);
```
Comment (rhvgoyal) [inline src/oci-umount.c]: same here.

File: src/oci-umount.c [hunk 4]
```diff
@@ -412,6 +422,119 @@ static int parent_mntid(const char *id, char *path, const struct mount_info *mnt
...
+		return 1;
+	}
+	pr_perror("%s: maximum number of configs used, can't process %s", id, file);
+	return -1;
+}
+
+static int load_config(const char *id, char **configs, const char *dirpath, struct host_mount_info *mounts_on_host, int *nr_umounts) {
+
+	/* Parse configuration dir for configuration file, canonicalize path
+	   names and skip paths which are not a mountpoint on host */
```
Comment (rhvgoyal) [inline src/oci-umount.c]: We don't require paths to be a mount point on host. So above comment is wrong?

File: src/oci-umount.c [hunk 5]
```diff
@@ -412,6 +422,119 @@ static int parent_mntid(const char *id, char *path, const struct mount_info *mnt
...
+	struct dirent *dp;
+	char path[PATH_MAX];
+	_cleanup_dir_ DIR *dir = opendir(dirpath);
+	if (!dir) {
+		pr_perror("%s: Failed to read directory %s", id, dirpath);
+		return EXIT_FAILURE;
+	}
+	while ((dp=readdir(dir)) != NULL) {
```
Comment (rhvgoyal) [inline src/oci-umount.c]: space around "=" ? There are other places too with such instances.

File: src/oci-umount.c [hunk 6]
```diff
@@ -412,6 +422,119 @@ static int parent_mntid(const char *id, char *path, const struct mount_info *mnt
...
+		mounts_on_host[nr_umounts].submounts_only = submounts_only;
+		nr_umounts++;
+	}
+
+	*number = nr_umounts;
+	return 0;
+}
+
+static int addconfig(const char *id, char **configs, const char *file) {
+	for (int i=0; i<MAX_CONFIGS; i++) {
```
Comment (rhvgoyal) [inline src/oci-umount.c]: spaces around = and <

File: src/oci-umount.c [hunk 7]
```diff
@@ -412,6 +422,119 @@ static int parent_mntid(const char *id, char *path, const struct mount_info *mnt
...
+		if (!configs[i]) {
+			configs[i]=strdup(file);
```
Comment (rhvgoyal) [inline src/oci-umount.c]: space round "="

File: src/oci-umount.c [hunk 8]
```diff
@@ -412,6 +422,119 @@ static int parent_mntid(const char *id, char *path, const struct mount_info *mnt
...
+	size_t len = 0;
+	int nr_umounts = *number;
+	/* Parse oci-umounts.conf file, canonicalize path names and skip
+	 * paths which are not a mountpoint on host */
+	fp = fopen(path, "r");
+	if (fp == NULL) {
+		if (errno == ENOENT) {
+			pr_pwarning("%s: Config file not found: %s", id, path);
+			return 0;
+		}
```
Comment (rhvgoyal) [inline src/oci-umount.c]: Now we call readconf() only if stat() returned success. That means we should never get ENOENT. So is this code still required?

File: src/oci-umount.c [hunk 9]
```diff
@@ -412,6 +422,119 @@ static int parent_mntid(const char *id, char *path, const struct mount_info *mnt
...
+			}
+		}
+
+		real_path = realpath(line, NULL);
+		if (!real_path) {
+			pr_pdebug("%s: Failed to canonicalize path [%s]: %m. Skipping.", id, line);
+			continue;
+		}
+
+		if (find_mount(real_path, mounts_on_host, nr_umounts))
```
Comment (rhvgoyal) [inline src/oci-umount.c]: Can we rename find_mount(). Its too generic and there are too many mounts floating around. How about `is_duplicate_path()` ?

**Comment by rhvgoyal** [comment]
Got following error message.

`umounthook <error>: cab0d2f222e0: Failed to read directory /usr/share/oci-umount/oci-umount.d: No such file or directory`

So is it mandatory to have this directory present? I would think no.

File: src/oci-umount.c (hunk 9 above)
Comment (rhvgoyal) [reply src/oci-umount.c]: Also we probably should drop a debug message that we are ignoring a path because it is a duplicate? It will not be a common situation so will not show up in logs by default. Will show up only if things are configured wrong. For example, when docker also introduces a config file of its own and it will have same entries in /etc/oci-umount.conf as well. That will be a good indicator to update oci-umount package and drop these entries from /etc/oci-umount.conf.

**Comment by rhvgoyal** [comment]
It does not seem to work if I just drop a file in /etc/oci-umount.d/oci-umount.conf. There is no /etc/oci-umount.conf and no /usr/share/oci-umount/oci-umount.d/

**Comment by rhatdan** [comment]
/etc/oci-umount/oci-umount.d is where it is looking.

**Comment by rhatdan** [comment]
/usr/share/oci-umount/oci-umount.d should be part of the package, so it should be there.  If you want it to be optional, then we can make are config optional, but you probably should not install the package then.

File: src/oci-umount.c (hunk 7 above)
Comment (rhatdan) [reply src/oci-umount.c]: Fixed

File: src/oci-umount.c [hunk 10]
```diff
@@ -412,6 +422,119 @@ static int parent_mntid(const char *id, char *path, const struct mount_info *mnt
...
+
+static int readconf(const char *id, const char *path, struct host_mount_info *mounts_on_host, int *number) {
+	_cleanup_fclose_ FILE *fp = NULL;
+	_cleanup_free_ char *line = NULL;
+	ssize_t read;
+	char *real_path;
+	size_t len = 0;
+	int nr_umounts = *number;
+	/* Parse oci-umounts.conf file, canonicalize path names and skip
+	 * paths which are not a mountpoint on host */
```
[the remark this answers is not in the document]
Comment (rhatdan) [reply src/oci-umount.c]: Fixed

File: src/oci-umount.c (hunk 8 above)
Comment (rhatdan) [reply src/oci-umount.c]: Removed

File: src/oci-umount.c (hunk 9 above)
Comment (rhatdan) [reply src/oci-umount.c]: Fixed

**Comment by rhatdan** [comment]
@rhvgoyal Everything is fixed except for failure on /usr/share/oci-umount/oci-umount.d directory.

**Comment by rhvgoyal** [comment]
@rhatdan I put oci-umount.conf in /etc/oci-umount.d/ and that's not working.

File: src/oci-umount.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (rhvgoyal) [reply src/oci-umount.c]: This comment is still not fixed? We are not requiring anything to be mount point on host.

**Comment by rhatdan** [comment]
@rhvgoyal 
I told you the path is 
/etc/oci-umount/oci-umount.d
NOT
/etc/oci-umount.d/
Put it in the right place and it should work.

```
 +#define ETCMOUNTDIR "/etc/oci-umount/oci-umount.d" 
```

File: src/oci-umount.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (rhatdan) [reply src/oci-umount.c]: I guess it should say path is not on the host.

[the remark this answers is not in the document]
Comment (rhatdan) [reply src/oci-umount.c]: Fixed

**Comment by rhvgoyal** [comment]
Hmm.., interesting. So why /etc/oci-umount/oci-umount.d and not /etc/oci-umount.d? I see so many other similar directories in /etc.

'/etc/sysctl.d/'
'/etc/yum.repos.d'
'/etc/xinet.d'
'/etc/popt.d'

And list goes on.

**Comment by rhvgoyal** [comment]
Or intent is to create a separate directory for oci-umount and then we can put more contents/dirs there in future?

**Comment by rhatdan** [comment]
My understanding of current packaging guidelines is to own the directory where you content is defined, not to scatter random directories/files all over /etc.

Like /etc/systemd rather then just thowing stuff in /etc.  It also matches up with /usr/share/oci-umount/oci-umount.d.

**Comment by rhatdan** [comment]
If we later want to have add additional config files we don't want to add /etc/oci-umount.json or other content.

**Comment by rhatdan** [comment]
@rhvgoyal This of this as just protecting us for future changes.  You have one directory named for package in /usr/share/PACKAG and one in /etc/PACKAGE.

**Comment by rhvgoyal** [comment]
LGTM

Thanks
